### PR TITLE
SOEOPSFY24-22

### PR DESCRIPTION
**Pull request by imonroe**
# READY FOR REVIEW

# Summary
- When you have a List paragraph with an Event list, and it's set in Card Grid view mode, and you limit the output to fewer than three (or if the query returns fewer than three) items, the items are squished to the left of the paragraph, with a hole for empty space.
- I'm proposing something like this CSS-only fix, but I'm not sure it's optimal.

# Review By (Date)
- When does this need to be reviewed by?

# Criticality
- 2

# Urgency
- normal

# Review Tasks

## Setup tasks and/or behavior to test

1. Check out this branch
2. Rebuild Cache
3. Create 3 events, dated in the future.
4. Create a page with a 2-column layout.  In the first column, add a List paragraph.  Set it to be an Event List, and choose card view.  For the limit, set it to 3.
5. observe the page, and note the card grid shows up correctly.
6. In the paragraph, change the limit to 2.
7. Observe the page, and note that only two items are shown, and they look right.
8. In the paragraph, change the limit to 1.  Note only one item shows up, and it looks right.

### Site Configuration Sync

- Is there a config:export in this PR that changes the config sync directory? NO.

## Front End Validation
- [ ] Design is approved by @ user?
- [ ] HTML validation: Is the markup using the appropriate semantic tags and [passes validation](https://validator.w3.org/nu/)? Or, [QA request ticket created](https://github.com/SU-SWS/template_warehouse/blob/master/jira_templates/QA_request_template.txt)?
- [ ] Cross-browser testing: Has been performed? Or, [QA request ticket created](https://github.com/SU-SWS/template_warehouse/blob/master/jira_templates/QA_request_template.txt)?
- [ ] Automated accessibility: Scans performed? Or, [QA request ticket created](https://github.com/SU-SWS/template_warehouse/blob/master/jira_templates/QA_request_template.txt)?
- [ ] Manual accessibility: Manually tested? Or, [QA request ticket created](https://github.com/SU-SWS/template_warehouse/blob/master/jira_templates/QA_request_template.txt)?

### Code
- [ ] Are the naming conventions following our standards?
- [ ] Does the code have sufficient inline comments?
- [ ] Is there anything in this code that would be hidden or hard to discover through the UI?
- [ ] Are there any [code smells](https://blog.codinghorror.com/code-smells/)?
- [ ] Are tests provided? eg (unit, behat, or codeception)

### Code security
- [ ] Are all [forms properly sanitized](https://www.drupal.org/docs/8/security/drupal-8-sanitizing-output)?
- [ ] Any obvious [security flaws or new areas for attack](https://www.drupal.org/docs/8/security)?

## General
- [ ] Is there anything included in this PR that is not related to the problem it is trying to solve?
- [ ] Is the approach to the problem appropriate?

# Affected Projects or Products
- Does this PR impact any particular projects, products, or modules?

# Associated Issues and/or People
- JIRA ticket(s)
- Other PRs
- Any other contextual information that might be helpful (e.g., description of a bug that this PR fixes, new functionality that it adds, etc.)
- Anyone who should be notified? (`@mention` them here)

# Resources
- [AMP Tool](https://stanford.levelaccess.net/index.php)
- [Accessibility Manual Test Script](https://docs.google.com/document/d/1ZXJ9RIUNXsS674ow9j3qJ2g1OAkCjmqMXl0Gs8XHEPQ/edit?usp=sharing)
- [HTML Validator](https://validator.w3.org/)
- [Browserstack](https://live.browserstack.com/dashboard) and link to [Browserstack Credentials](https://asconfluence.stanford.edu/confluence/display/SWS/External+Account+Credentials)

### CODE REVIEW

**Comment by imonroe** [comment]
Here's how it should look with a three item limit: 
image

**Comment by imonroe** [comment]
Here's how it looks with a limit of two: 

image

**Comment by imonroe** [comment]
Here's how it looks with a limit of one: 
image

**Comment by imonroe** [comment]
Here is what it looks like with a limit set to one: 

image

Here's what it looks like with a limit set to two: 

image

Here's what it looks like with a limit set to three: 

image

**Comment by imonroe** [comment]
@jenbreese Does this seem like a good solution to you?  For context, see: https://stanfordits.atlassian.net/browse/SOEOPSFY24-22

File: themes/soe_basic/src/scss/theme/_events.scss [hunk 1]
```diff
@@ -289,3 +289,16 @@
...
+
+// Event List paragraph, card view.
+.view.stanford-events.cards {
+  .su-list-unstyled.grid-container-3 {
+    display: flex;
+    justify-content: flex-start;
+    gap: 1em;
+
+    li {
+      flex: 1 0 29;
```
Comment (jenbreese) [inline themes/soe_basic/src/scss/theme/_events.scss]: @imonroe Can you explain what you want to do here? This isn't valid CSS. Do you want all the cards evenly sized? 
Screenshot 2023-12-13 at 9 58 13 AM

Comment (jenbreese) [reply themes/soe_basic/src/scss/theme/_events.scss]: I think this is better: `flex: 1 1 min-content;`
It is evenly divided if it is 1, 2 or 3 cards.

Comment (imonroe) [reply themes/soe_basic/src/scss/theme/_events.scss]: Ok, I'll try that.  FWIW, that is supposed to mean: 
- flex-grow is ok
- flex-shrink is not ok
- flex-basis is initially 29%.  I did forget the percentage there, which is why it's not valid.

Comment (jenbreese) [reply themes/soe_basic/src/scss/theme/_events.scss]: Thanks!

Comment (imonroe) [reply themes/soe_basic/src/scss/theme/_events.scss]: Your suggestion works great, Jen.  I've made the change.

**Comment by pookmish** [comment]
There is a trade off with this approach. If you choose to display more than 3 but only 4 items are available, the 4th item takes up the whole area. I think this is why we've been using grid more to keep them all the same regardless.

You could do some JS/template logic that adds/removes classes based on the number of elements. ie  1-3 use the flex, 4+ use grid. :shrug:

**Comment by imonroe** [comment]
Pinging @pookmish on this, as it sort of fell off my radar.  Do we really want to build template logic as suggested above?  My instinct is that building complex template logic to deal with numbers of items divisible by four is maybe overkill?  The original ticket was an edge case, and having three columns and four items is another edge case.  I'll rework it if necessary, but my feeling is that it might be overkill....

**Comment by imonroe** [comment]
It did need to have a `wrap` added.

This is with five items.

image

This is with four items.  After six, it wraps again.

image

**Comment by pookmish** [comment]
Yeah, so that's the tradeoff with the `flex`. The 4th items with 4 total, 7th item with 7 total, etc, take up the whole space. Some think it's acceptable some don't like it. I'm personally in the latter because i don't like one event card to be more space consuming than others. But it's really up to the client/UX to make that choice. The actual approach in the PR is acceptable to me.

**Comment by imonroe** [comment]
Well, @buttonwillowsix made the request originally, so I'll see if she'd like to review before merging.

**Comment by buttonwillowsix** [comment]
This looks great to me!

**Comment by pookmish** [comment]
Good with her, good with me. Probably need to rebase/merge the lastest `11.x` into this branch though

**Pull request merged**